reqwest-middleware/src/req_init.rs

```diff
@@ -0,0 +1,84 @@
+use crate::RequestBuilder;
+
+/// When attached to a [`ClientWithMiddleware`] (generally using [`with_init`]), it is run
+/// whenever the client starts building a request, in the order it was attached.
+///
+/// # Example
+///
+/// ```
+/// use reqwest_middleware::{RequestInitialiser, RequestBuilder};
+///
+/// struct AuthInit;
+///
+/// impl RequestInitialiser for AuthInit {
+///     fn init(&self, req: RequestBuilder) -> RequestBuilder {
+///         req.bearer_auth("my_auth_token")
+///     }
+/// }
+/// ```
+///
+/// [`ClientWithMiddleware`]: crate::ClientWithMiddleware
+/// [`with_init`]: crate::ClientBuilder::with_init
+pub trait RequestInitialiser: 'static + Send + Sync {
+    fn init(&self, req: RequestBuilder) -> RequestBuilder;
+}
+
+impl<F> RequestInitialiser for F
+where
+    F: Send + Sync + 'static + Fn(RequestBuilder) -> RequestBuilder,
+{
+    fn init(&self, req: RequestBuilder) -> RequestBuilder {
+        (self)(req)
+    }
+}
+
+/// A middleware that inserts the value into the [`Extensions`](task_local_extensions::Extensions) during the call.
+///
+/// This is a good way to inject extensions to middleware deeper in the stack
+///
+/// ```
+/// use reqwest::{Client, Request, Response};
+/// use reqwest_middleware::{ClientBuilder, Middleware, Next, Result, Extension};
+/// use task_local_extensions::Extensions;
+///
+/// #[derive(Clone)]
+/// struct LogName(&'static str);
+/// struct LoggingMiddleware;
+///
+/// #[async_trait::async_trait]
+/// impl Middleware for LoggingMiddleware {
+///     async fn handle(
+///         &self,
+///         req: Request,
+///         extensions: &mut Extensions,
+///         next: Next<'_>,
+///     ) -> Result<Response> {
+///         // get the log name or default to "unknown"
+///         let name = extensions
+///             .get()
+///             .map(|&LogName(name)| name)
+///             .unwrap_or("unknown");
+///         println!("[{name}] Request started {req:?}");
+///         let res = next.run(req, extensions).await;
+///         println!("[{name}] Result: {res:?}");
+///         res
+///     }
+/// }
+///
+/// async fn run() {
+///     let reqwest_client = Client::builder().build().unwrap();
+///     let client = ClientBuilder::new(reqwest_client)
+///         .with_init(Extension(LogName("my-client")))
+///         .with(LoggingMiddleware)
+///         .build();
```

Comment:
In this example couldn't the `LogName` have been baked into `LoggingMiddleware`. Or if the latter was a shared thing, a wrapper middleware used. I'm not seeing the motivating case for the new abstractions from this example very clearly.

Reply:
Ah, I thought I added the extra example. The case is where you might have a client wide extension, and then you can have a per-request configuration too.

```rust
let client = ClientBuilder::new(reqwest_client)
    .with_init(Extension(LogName("my-client")))
    .with(LoggingMiddleware)
    .build();
    
// uses "my-client" as the log name
client.get("https://truelayer.com").send().await

// uses "create-payment" as the log name
client.post("https://api.truelayer.com/payment")
    .with_extension(LogName("create-payment"))
    .send()
    .await
```

Reply:
And this is something I want to supplement #54 because I'm not keen on breaking the current `SpanBackend` trait to add `&self` just yet, but it does have access to the extensions.

The desired configuration by the author was to have client wide configuration of the span name, but I think a request specific span name might be more desirable in some situations. So this is the best of both worlds for configuration